Automate calls to the SonarQube static code analysis tool in the nightly build

*Replace italics below with details for this issue.*

## Describe the New Feature ##
Add script to run SonarQube scanner

### Acceptance Testing ###
*List input data types and sources.*
*Describe tests required for new functionality.*

Create the scripts of SonarQube and add to the crontab for the night build under "met_test" account.

### Time Estimate ###
*Estimate the amount of work required here.*
4 hours

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
2771021

## Define the Metadata ##

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [ ] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## New Feature Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

Build problem at src/basic/vx_math on seneca:
hsoh@seneca:/d1/personal/hsoh/git/release/MET_tag_v10.1.0-beta5/met/src/basic/vx_math$ make
make: *** No rule to make target 'viewgravity_to_string.cc', needed by 'libvx_math_a-viewgravity_to_string.o'.  Stop.


at kiowa:
/d1/personal/hsoh/git/pull_request/MET_feature_1844_python_embedding/met/src/basic/enum_to_string/enum_to_string -concat_string affine.h

enum_to_string:  Making header file "viewgravity_to_string.h"
enum_to_string:  Making source file "viewgravity_to_string.cc"
g++ -DHAVE_CONFIG_H -I. -I../../..  -I../../../src/basic/vx_cal -I../../../src/basic/vx_config -I../../../src/basic/vx_log -I../../../src/basic/vx_math -I../../../src/basic/vx_util -I../../../src/libcode/vx_afm -I../../../src/libcode/vx_analysis_util -I../../../src/libcode/vx_color -I../../../src/libcode/vx_data2d -I../../../src/libcode/vx_data2d_factory -I../../../src/libcode/vx_data2d_grib -I../../../src/libcode/vx_data2d_grib2 -I../../../src/libcode/vx_data2d_nccf -I../../../src/libcode/vx_data2d_nc_met -I../../../src/libcode/vx_data2d_nc_pinterp -I../../../src/libcode/vx_data2d_python -I../../../src/libcode/vx_python3_utils -I../../../src/libcode/vx_geodesy -I../../../src/libcode/vx_gis -I../../../src/libcode/vx_gnomon -I../../../src/libcode/vx_grid -I../../../src/libcode/vx_gsl_prob -I../../../src/libcode/vx_nav -I../../../src/libcode/vx_nc_obs -I../../../src/libcode/vx_nc_util -I../../../src/libcode/vx_pb_util -I../../../src/libcode/vx_plot_util -I../../../src/libcode/vx_pointdata_python -I../../../src/libcode/vx_ps -I../../../src/libcode/vx_pxm -I../../../src/libcode/vx_render -I../../../src/libcode/vx_regrid -I../../../src/libcode/vx_shapedata -I../../../src/libcode/vx_solar -I../../../src/libcode/vx_statistics -I../../../src/libcode/vx_stat_out -I../../../src/libcode/vx_summary -I../../../src/libcode/vx_time_series -I../../../src/libcode/vx_series_data -I../../../src/libcode/vx_tc_util -fopenmp -I/usr/local/netcdf-4.7.0/gcc-6.3.0/include -I/usr/local/hdf5-1.8.20/include -I/usr/local/include -I../../basic/vx_log -I../../basic/vx_util -I/var/autofs/mnt/linux-amd64/debian/stretch/local/met-python3/include/python3.7m -I/d1/projects/MET/MET_releases/external_libs/gnu_6.3.0/include -I/usr/local/hdf4-4.2.13/include/hdf -I/usr/local/hdf-eos2-19v1/include -I/usr/include/cairo -I/usr/include/freetype2 -DBLOCK4 -DWITH_GRIB2 -DWITH_PYTHON -DMET_BASE="\"/d1/personal/hsoh/git/pull_request/MET_feature_1844_python_embedding/met/share/met\""  -g -DUNDERSCORE -fPIC -D__64BIT__ -MT libvx_math_a-viewgravity_to_string.o -MD -MP -MF .deps/libvx_math_a-viewgravity_to_string.Tpo -c -o libvx_math_a-viewgravity_to_string.o `test -f 'viewgravity_to_string.cc' || echo './'`viewgravity_to_string.cc
anlib libvx_math.a

--- a/scripts/environment/development.seneca
+++ b/scripts/environment/development.seneca
@@ -49,3 +49,8 @@ export MET_TEST_RSCRIPT=/usr/local/R-4.1.2/bin/Rscript
 export PATH="/usr/local/nco/bin:/usr/local/netcdf/bin:\
              /usr/local/sbin:/usr/local/bin:/usr/sbin:\
              /usr/bin:/sbin:/bin:/usr/bin/X11:/opt/bin:$PATH"
+
+# SonarQube
+export SONARQUBE_DIR=/d1/projects/SonarQube/
+export SONARQUBE_WRAPPER_BIN=$SONARQUBE_DIR/build-wrapper-linux-x86
+export SONARQUBE_SCANNER_BIN=$SONARQUBE_DIR/sonar-scanner-4.6.2.2472-linux/bin

--- a/scripts/sonarqube/run_nightly.sh
+++ b/scripts/sonarqube/run_nightly.sh
@@ -1,0 +1,75 @@
+#!/bin/bash
+#
+# Run nightly SonarQube scan
+#=======================================================================
+#
+# This run_nightly.sh script calls the run_sonarqube.sh script.
+# It is intented to be run nightly through cron. Output should be
+# directed to the LOGFILE, per cron convention. To run this script, use
+# the following commands:
+#
+#    git clone https://github.com/dtcenter/MET
+#    MET/scripts/sosnarqube/run_nightly.sh name
+#
+# Usage: run_nightly.sh name
+#    where "name" specifies a branch, tag, or hash
+#
+# For example, scan the develop branch:
+#    run_nightly.sh develop
+#
+#=======================================================================
+
+# Constants
+#EMAIL_LIST="johnhg@ucar.edu hsoh@ucar.edu jpresto@ucar.edu linden@ucar.edu mccabe@ucar.edu"
+EMAIL_LIST="hsoh@ucar.edu"
+KEEP_DAYS=5
+
+function usage {
+  echo
+  echo "USAGE: run_nightly.sh name"
+  echo "   where \"name\" specifies a branch, tag, or hash."
+  echo
+}
+
+# Check for arguments
+if [ $# -lt 1 ]; then usage; exit 1; fi
+
+# Store the full path to the scripts directory
+SCRIPT_DIR=`dirname $0`
+if [[ ${0:0:1} != "/" ]]; then SCRIPT_DIR=$(pwd)/${SCRIPT_DIR}; fi 
+
+# Define the development environment
+ENV_FILE=${SCRIPT_DIR}/../environment/development.`hostname`
+if [[ ! -e ${ENV_FILE} ]]; then
+  echo "$0: ERROR -> Development environment file missing: ${ENV_FILE}"
+  exit 1
+fi
+source ${ENV_FILE}
+
+# Delete old directories
+find ${MET_PROJ_DIR}/MET_regression/sonarqube \
+     -mtime +${KEEP_DAYS} -name "NB*" | \
+     xargs rm -rf
+
+# Create and switch to a run directory
+TODAY=`date +%Y%m%d`
+YESTERDAY=`date -d "1 day ago" +%Y%m%d`
+RUN_DIR=${MET_PROJ_DIR}/MET_regression/sonarqube/NB${TODAY}
+if [[ -e ${RUN_DIR} ]]; then rm -rf ${RUN_DIR}; fi
+mkdir -p ${RUN_DIR}
+cd ${RUN_DIR}
+
+# Create a logfile
+LOGFILE=${RUN_DIR}/run_sonarqube_${TODAY}.log
+
+# Run scan and check for bad return status
+${SCRIPT_DIR}/run_sonarqube.sh ${1} > ${LOGFILE}
+if [[ $? -ne 0 ]]; then
+  echo "$0: The nightly SonarQube scan FAILED in `basename ${RUN_DIR}`." >> ${LOGFILE}
+  cat ${LOGFILE} | mail -s "MET SonarQube scan Failed for ${1} in `basename ${RUN_DIR}` (autogen msg)" ${EMAIL_LIST}
+  exit 1
+fi
+
+# Convert SonarQube report from pdf to html
+
+exit 0

--- a/scripts/sonarqube/run_sonarqube.sh
+++ b/scripts/sonarqube/run_sonarqube.sh
@@ -1,0 +1,144 @@
+#!/bin/bash
+#
+# Run SonarQube Source Code Analyzer on a specified revision of MET
+#=======================================================================
+#
+# This run_sonarqube.sh script will check out the specified version
+# of MET and run the SonarQube Source Code Analyzer on it.  First,
+# go to the directory where you would like the SCA output written and
+# then run:
+#
+#    git clone https://github.com/dtcenter/MET
+#    MET/scripts/sonarqube/run_sonarqube_sca.sh name
+#
+# Usage: run_sonarqube_sca.sh name
+#    Test the specified branched version of MET:
+#       run_sonarqube_sca.sh {branch name}
+#    Test the specified tagged version of MET:
+#       run_sonarqube_sca.sh {tag name}
+#
+#=======================================================================
+
+# Constants
+GIT_REPO="https://github.com/dtcenter/MET"
+
+function usage {
+        echo
+        echo "USAGE: $(basename $0) name"
+        echo "   where \"name\" specifies a branch, tag, or hash."
+        echo
+}
+
+# Check for arguments
+if [[ $# -lt 1 ]]; then usage; exit; fi
+
+# Check that SONARQUBE_WRAPPER_BIN is defined
+if [ -z ${SONARQUBE_WRAPPER_BIN} ]; then
+  which build-wrapper-linux-x86-64 2> /dev/null
+  if [ $? -eq 0 ]; then
+    SONARQUBE_WRAPPER_BIN=$(which build-wrapper-linux-x86-64 2> /dev/null)
+  else
+    which build-wrapper 2> /dev/null
+    if [ $? -eq 0 ]; then
+      SONARQUBE_WRAPPER_BIN=$(which build-wrapper 2> /dev/null)
+    else
+      echo "ERROR: SONARQUBE_WRAPPER_BIN must be set"
+      exit 1
+    fi
+  fi
+fi
+if [ ! -e ${SONARQUBE_WRAPPER_BIN} ]; then
+  echo "ERROR: SONARQUBE_WRAPPER_BIN (${SONARQUBE_WRAPPER_BIN}) does not exist"
+  exit 1
+fi
+
+# Check that SONARQUBE_SCANNER_BIN is defined
+if [ -z ${SONARQUBE_SCANNER_BIN} ]; then
+  which sonar-scanner 2> /dev/null
+  if [ $? -eq 0 ]; then
+    SONARQUBE_SCANNER_BIN=$(which sonar-scanner 2> /dev/null)
+  else
+    echo "ERROR: SONARQUBE_SCANNER_BIN must be set"
+    exit 1
+  fi
+fi
+if [ ! -e ${SONARQUBE_SCANNER_BIN} ]; then
+  echo "ERROR: SONARQUBE_SCANNER_BIN (${SONARQUBE_SCANNER_BIN}) does not exist"
+  exit 1
+fi
+
+if [ -z ${SONARQUBE_OUT_DIR} ]; then
+  export SONARQUBE_OUT_DIR=bw-outputs
+fi
+
+# Sub-routine for running a command and checking return status
+function run_command() {
+
+  # Print the command being called
+  echo "CALLING: $1"
+
+  # Run the command and store the return status
+  $1
+  STATUS=$?
+
+  # Check return status
+  if [[ ${STATUS} -ne 0 ]]; then
+     echo "ERROR: Command returned with non-zero status ($STATUS): $1"
+     exit ${STATUS}
+  fi
+
+  return ${STATUS}
+}
+
+
+# Store the full path to the scripts directory
+SCRIPT_DIR=`dirname $0`
+if [[ ${0:0:1} != "/" ]]; then SCRIPT_DIR=$(pwd)/${SCRIPT_DIR}; fi 
+
+# Clone repo into a sub-directory and checkout the requested version
+REPO_DIR="MET-${1}"
+
+if [ -e ${REPO_DIR} ]; then
+  run_command "rm -rf ${REPO_DIR}"
+fi
+run_command "git clone ${GIT_REPO} ${REPO_DIR}"
+run_command "cd ${REPO_DIR}"
+run_command "git checkout ${1}"
+
+# Build the MET instance
+run_command "cd met"
+
+# Run bootstrap
+run_command "./bootstrap"
+
+# Do no manually set the CXX and F77 compilers.
+# Let the configure script pick them.
+# Otherwise, the SonarQube logic does not work.
+export MET_DEVELOPMENT=true
+
+# Run the configure script
+run_command "./configure --prefix=`pwd` \
+            --enable-grib2 \
+            --enable-modis \
+            --enable-mode_graphics \
+            --enable-lidar2nc \
+            --enable-python"
+
+# Set the build id
+#BUILD_ID="MET-${1}"
+
+# Copy sonar-project.properties
+[ ! -e "sonar-project.properties" ] && cp -p $SCRIPT_DIR/sonar-project.properties .
+
+# Run SonarQube clean
+run_command "make clean"
+
+
+# Run SonarQube make
+run_command "${SONARQUBE_WRAPPER_BIN}/build-wrapper-linux-x86-64 --out-dir $SONARQUBE_OUT_DIR make"
+
+# Run SonarQube scan
+run_command "${SONARQUBE_SCANNER_BIN}/sonar-scanner"
+
+# Run SonarQube report generator to make a PDF file
+#TODAY=`date +%Y%m%d`

--- a/scripts/sonarqube/sonar-project.properties
+++ b/scripts/sonarqube/sonar-project.properties
@@ -1,0 +1,18 @@
+sonar.projectKey=org.sonarqube:MET_develop_NB
+sonar.projectName=MET Nightly build
+sonar.projectVersion=1.0
+
+sonar.sources=src
+
+# The build-wrapper output dir
+sonar.cfamily.build-wrapper-output=bw-outputs
+
+# Encoding of the source files
+sonar.sourceEncoding=UTF-8
+
+#----- Default SonarQube server
+#sonar.host.url=http://localhost:9000
+sonar.host.url=http://mandan:9000
+
+sonar.login=met
+sonar.password=met@sonar.ucar

